File: packages/runtime/src/artifact.ts
import * as tg from "./index.ts";

export type Artifact = tg.Directory | tg.File | tg.Symlink;

export namespace Artifact {
	export type Id = string;

	export let withId = (id: Artifact.Id): Artifact => {
		let prefix = id.substring(0, 3);
		if (prefix === "dir") {
			return tg.Directory.withId(id);
		} else if (prefix === "fil") {
			return tg.File.withId(id);
		} else if (prefix === "sym") {
			return tg.Symlink.withId(id);
		} else {
			throw new Error(`invalid artifact id: ${id}`);
		}
	};

	export let is = (value: unknown): value is Artifact => {
		return (
			value instanceof tg.Directory ||
			value instanceof tg.File ||
			value instanceof tg.Symlink
		);
	};

	export let expect = (value: unknown): Artifact => {
		tg.assert(is(value));
		return value;
	};

	export let assert = (value: unknown): asserts value is Artifact => {
		tg.assert(is(value));
	};
}

export let isGraphArg = (
	obj: unknown,
): obj is { graph: tg.Graph; node: number } => {
	return (
		typeof obj === "object" &&
		obj !== null &&
		"graph" in obj &&
		obj.graph instanceof tg.Graph &&
		"node" in obj &&
		typeof obj.node === "number"
	);
};

File: packages/clients/python/tangram/artifact.py
"""Artifact type and utilities.

This module mirrors the JavaScript packages/clients/js/src/artifact.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, Union

if TYPE_CHECKING:
    from tangram.directory import Directory
    from tangram.file import File
    from tangram.symlink import Symlink

# Artifact is a union of Directory, File, and Symlink.
Artifact = Union["Directory", "File", "Symlink"]

# Type aliases.
Id = str
Kind = Literal["directory", "file", "symlink"]


def with_id(id: Id) -> Artifact:
    """Create an artifact from its ID."""
    from tangram.assert_ import assert_
    from tangram.directory import Directory
    from tangram.file import File
    from tangram.symlink import Symlink

    assert_(isinstance(id, str), f"expected a string: {id}")
    prefix = id[:3] if len(id) >= 3 else ""
    if prefix == "dir":
        return Directory.with_id(id)
    elif prefix == "fil":
        return File.with_id(id)
    elif prefix == "sym":
        return Symlink.with_id(id)
    else:
        raise ValueError(f"invalid artifact id: {id}")


def is_(value: Any) -> bool:
    """Check if a value is an Artifact."""
    from tangram.directory import Directory
    from tangram.file import File
    from tangram.symlink import Symlink

    return isinstance(value, (Directory, File, Symlink))


def expect(value: Any) -> Artifact:
    """Assert that a value is an Artifact and return it."""
    from tangram.assert_ import assert_

    assert_(is_(value))
    return value


def assert_(value: Any) -> None:
    """Assert that a value is an Artifact."""
    from tangram.assert_ import assert_ as assert_fn

    assert_fn(is_(value))
